### ISSUE TITLE
Corporate netsec software on Windows may causes 'self signed certificate' verification errors

### ISSUE THREAD

**Issue by aaronsteers** [meltano/meltano#6334]
Two instances so far reported in slack:

- https://meltano.slack.com/archives/C01TCRBBJD7/p1656621387384619
- https://meltano.slack.com/archives/C01TCRBBJD7/p1656516344303779

**Comment by aaronsteers** [comment]
> From: [@Greg Vaslo](https://meltano.slack.com/team/U032EQKNP1P)
>
> On Linux it's relatively easy to manage custom Root CA Cert Stores for Python request  via the REQUESTS_CA_BUNDLE environment variable. Depending on your Linux distribution, your custom Root CA Cert store will be in different locations, but basically, as long as you set the REQUESTS_CA_BUNDLE to the path that contains to the PEM file (e.g. /etc/ssl/certs/ca-certificates.crt on Debian/Ubuntu which is managed by update-ca-certificates to merge your internal custom Enterprise Root CAs) that has your complete customized Root CA bundle, then at least the majority of Python requests calls will work (Meltano SDK plugins won't work though, until we can do some more testing and work on this PR: https://github.com/meltano/sdk/issues/633)
>
> With Python on Windows, the REQUESTS_CA_BUNDLE variable will still work, but on Windows you don't get the luxury of having the OS provide you a convenient file location where Windows would store the custom Root CA Cert store bundle for you in PEM format. You will need to generate such a file on Windows yourself (which is possible with certutil to dump your custom Windows Root CA Cert stores into a PEM formatted file)
>

**Comment by visch** [comment]
Looks like https://stackoverflow.com/questions/42981429/ssl-failure-on-windows-using-python-requests
Specefically

```
import requests
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.ssl_ import create_urllib3_context

class SSLContextAdapter(HTTPAdapter):
    def init_poolmanager(self, *args, **kwargs):
        context = create_urllib3_context()
        kwargs['ssl_context'] = context
        context.load_default_certs() # this loads the OS defaults on Windows
        return super(SSLContextAdapter, self).init_poolmanager(*args, **kwargs)

s = requests.Session()
adapter = SSLContextAdapter()
s.mount('https://myinternalsite', adapter)
response = s.get('https://myinternalsite')
```

Will load your Windows certs automatically for you (from your machine). I'd assume in a corp environment those certs are already on the persons machine. The fact that the default with Python is to use it's own root store is a curious choice but 🤷

**Comment by Vaslo** [comment]
All, I've come back again to look at this issue.  I decided to try a few environments from the same PC.  I started with the last version of Meltano v1 (1.105.0).  I was able to install this and had no issues loading the taps and targets I needed.  If I install a fresh environment with 2.0 or beyond, I run into the same issue as mentioned above.  Is this because of the change to pull directly from Meltano hub?  Is there a way to go back to pulling wherever it was pulling from before 2.0?  I'm not getting much help from corporate IT.  I'll also check in the Slack and see if there are any final ideas and will try just starting in 1.105.0 and then upgrading after I install the extractor and loader.

**Comment by visch** [comment]
Discovery file changed locations, maybe a workaround you put in place is broken now? Maybe you could take a look at  lock files which is new https://docs.meltano.com/reference/command-line-interface#lock I believe, then meltano shouldn't need to be able to reach the hub. I am guessing at what the cause is without errors and things but this sounds likely based on the past info!

**Comment by edgarrmondragon** [comment]
A workaround for anyone stumbling into this:

1. Locate the plugin in Meltano Hub: https://hub.meltano.com/. In the plugin page, take note of the default _variant_, or pick the one you prefer:

   <img width="785" alt="Screen Shot 2022-11-07 at 17 40 46" src="https://user-images.githubusercontent.com/16805946/200438731-a78b4c3a-e4fa-4322-a644-5870d7018130.png">

1. Manually download the plugin definition from the Meltano Hub API into the `plugins/<plugin type>` directory of your Meltano project. In the case of `tap-oracle`, this directory is `plugins/extractors/`:

   ```bash
   PLUGIN_TYPE=extractors
   PLUGIN=tap-oracle
   VARIANT=transferwise
   PLUGIN_PATH=plugins/$PLUGIN_TYPE/$PLUGIN--$VARIANT
   curl https://hub.meltano.com/meltano/api/v1/$PLUGIN_PATH > $PLUGIN_PATH.lock
   ```

1. Add the plugin entry to `meltano.yml`

   ```yaml
   plugins:
     extractors:
     - name: tap-oracle
       variant: transferwise
   ```

You should now be able to use the plugin normally.

**Comment by WillDaSilva** [comment]
Still relevant

**Comment by gvas7** [comment]
> A workaround to this error in `meltano add` for anyone stumbling into it:
> 
> 1. Locate the plugin in Meltano Hub: https://hub.meltano.com/. In the plugin page, take note of the default _variant_, or pick the one you prefer:
>    <img alt="Screen Shot 2022-11-07 at 17 40 46" width="785" src="https://user-images.githubusercontent.com/16805946/200438731-a78b4c3a-e4fa-4322-a644-5870d7018130.png">
> 2. Manually download the plugin definition from the Meltano Hub API into the `plugins/<plugin type>` directory of your Meltano project. In the case of `tap-oracle`, this directory is `plugins/extractors/`:
>    ```shell
>    PLUGIN_TYPE=extractors
>    PLUGIN=tap-oracle
>    VARIANT=transferwise
>    PLUGIN_PATH=plugins/$PLUGIN_TYPE/$PLUGIN--$VARIANT
>    curl https://hub.meltano.com/meltano/api/v1/$PLUGIN_PATH > $PLUGIN_PATH.lock
>    ```
> 3. Add the plugin entry to `meltano.yml`
>    ```yaml
>    plugins:
>      extractors:
>      - name: tap-oracle
>        variant: transferwise
>    ```
> 
> You should now be able to use the plugin normally.

It's been a long time and I've come back to try Meltano again since I now have access to some better data sources like Snowflake.  I want to try it again but still having the persistent certificate issues.  I tried your workaround here but it's saying "Block tap-csv not found".  I tried this both your way in WSL, and I also translated your method to Powershell where I also work and it just will not recognize the tap when I `meltano run` or `meltano invoke`.  To be clear:

1.) I use the 5 lines you designated but have changed it to `tap-csv` and variant `meltanolabs`. 
2.) It download the `.lock` file into the meltano logic under the folder `/snow_to_sql/plugins/extractors/` where `snow_to_sql` is the project folder.
3.) The downloaded file appears to have some kind of dictionary that starts with `{"capabilities":["catalog", "discover"],...`

Is there a missing step here?  Is there something I need to do to have the tap recognized by Meltano?  Do I need to also manually download the repo into the same folder?

Thanks.

**Comment by edgarrmondragon** [comment]
@gvas7 are you on Meltano 3.0+?

**Comment by gvas7** [comment]
> @gvas7 are you on Meltano 3.0+?

Yes I am - is this the wrong approach for 3?  I realize this topic is pretty dated.

**Comment by edgarrmondragon** [comment]
The easiest workaround for you then might be using [`--from-ref`](https://docs.meltano.com/reference/command-line-interface/#how-to-use):

```
meltano add extractor tap-csv --from-ref https://raw.githubusercontent.com/meltano/hub/main/_data/meltano/extractors/tap-csv/meltanolabs.yml
```
